PETWorks/profitability.py

```diff
@@ -0,0 +1,238 @@
+
+    qiIndices = _getQiIndices(anonymizedSubset.getHandle())
+
+    sampleRowIndex = -1
+    allSuppressed = False
+    for i in range(subsetRowNum):
+        for j in range(subsetColNum):
+            if not (j in qiIndices):
+                continue
+
+            if subsetDataFrame.iloc[i][j] != "*":
+                sampleRowIndex = i
+                break
+
+        if sampleRowIndex != -1:
+            break
+
+        allSuppressed = i == subsetRowNum - 1
+
+    anonymousLevels = []
+    for index in range(subsetColNum):
+        if not (index in qiIndices):
+            continue
+
+        value = subsetDataFrame.iloc[sampleRowIndex][index]
+        attributeName = subsetDataFrame.columns[index]
+        hierarchy = hierarchies[attributeName].getHierarchy()
+
+        if allSuppressed:
+            anonymousLevels.append(len(hierarchy[index])-1)
+            continue
+
+        anonymousLevels.append(_findAnonymousLevel(hierarchy, value))
+
+    return anonymousLevels
+
+
+def _getDataFrame(dataHandle: str) -> pd.DataFrame:
+    rowNum = dataHandle.getNumRows()
+    colNum = dataHandle.getNumColumns()
+
+    colNames = []
+    data = []
+    for i in range(rowNum):
+        row = []
+        for j in range(colNum):
+            row.append(dataHandle.getValue(i, j))
+
+            if i == 0:
+                colNames.append(dataHandle.getAttributeName(j))
+
+        data.append(row)
+
+    return pd.DataFrame(data, columns=colNames)
+
+
+def _getSubsetIndices(
+    table: str, subset: str,
+    hierarchies: dict[str, list[list[str]]],
+) -> list[int]:
+
+    qiNames = table.getDefinition().getQuasiIdentifyingAttributes()
+    qis = [qi for qi in qiNames]
+    qiIndices = _getQiIndices(table)
+
+    tableDataFrame = _getDataFrame(table)
+    groupedSubset = _getDataFrame(subset).groupby(qis)
+
+    tableRowNum = len(tableDataFrame)
+
+    subsetIndices = []
+    for _, subsetGroup in groupedSubset:
+
+        subsetGroupList = subsetGroup.values.tolist()
+        filter = pd.Series(True, index=range(tableRowNum))
+        for i in range(len(qiIndices)):
+            qiName = qis[i]
+            qiIndex = qiIndices[i]
+            filter &= tableDataFrame[qiName] == subsetGroupList[0][qiIndex]
+
+        subsetIndices += np.flatnonzero(filter).tolist()[:len(subsetGroupList)]
+
+    return subsetIndices
+
+
+def _getAnonymousData(
+        original: Data, anonymousLevels: list[int]
+) -> str:
+    levels = gateway.new_array(Int, len(anonymousLevels))
+    for i in range(len(anonymousLevels)):
+        levels[i] = anonymousLevels[i]
+
+    arxconfig = ARXConfiguration.create()
+    arxconfig.addPrivacyModel(KAnonymity(1))
+    anonymizer = ARXAnonymizer()
+    result = anonymizer.anonymize(original, arxconfig)
+
+    lattice = result.getLattice()
+    node = lattice.getNode(levels)
+
+    return result.getOutput(node, True)
+
+
+def _measureProfitability(
+    original: Data, subsetIndices: list[int],
```

Comment:
Please insert a line break after `original: Data,` and make the argument `subsetIndices: list[int],` easy to recognize. Beside, I suggest using tool `black` to format the code rather than doing it manually.